cdap-formats/src/main/java/io/cdap/cdap/format/io/JsonStructuredRecordDatumReader.java

```diff
@@ -149,12 +167,28 @@ protected Object decodeUnion(Decoder decoder, Schema unionSchema) throws IOExcep
       if (SCHEMA_TO_JSON_TYPE.get(schema.getType()) == token) {
         return decode(decoder, schema);
       }
+      if (LOGICAL_SCHEMA_TO_JSON_TYPE.get(schema.getLogicalType()) == token) {
+        return decode(decoder, schema);
+      }
     }
 
     throw new IOException(String.format("No matching schema found for union type: %s for token: %s", unionSchema,
                                         token));
   }
 
+  protected byte[] decodeDecimal(Decoder decoder, Schema decimalSchema) throws IOException {
+    JsonReader jsonReader = getJsonReader(decoder);
+    String strVal = jsonReader.nextString();
+    return FormatUtils.parseDecimal(decimalSchema, strVal).unscaledValue().toByteArray();
+  }
+
+  @Override
+  protected Object decodeBytes(Decoder decoder, Schema schema) throws IOException {
```

Comment:
Why don't you want to put it into JsonReader itself? This way you don't need special logic here. Also it's better to support reading both formats to preserve backwards compatibility

Reply:
Also please update PR header to say that it changes byte array handling as well